<a href="https://colab.research.google.com/github/SandroP10/Sales-Forecast-Project/blob/main/Regression_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import os
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import OrdinalEncoder
from scipy.stats import mode
from sklearn.metrics import mean_squared_error, mean_absolute_percentage_error

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/ca_training_data_hs23/final_feat.csv')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15567 entries, 0 to 15566
Data columns (total 21 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   article_name        15567 non-null  float64
 1   date_id             15567 non-null  int64  
 2   year                15567 non-null  int64  
 3   month               15567 non-null  int64  
 4   week                15567 non-null  int64  
 5   boss_thema_id       15567 non-null  int64  
 6   promo_01            15567 non-null  int64  
 7   promo_02            15567 non-null  int64  
 8   promo_03            15567 non-null  int64  
 9   promo_04            15567 non-null  int64  
 10  promo_05            15567 non-null  int64  
 11  sales               15567 non-null  float64
 12  zscore              15567 non-null  float64
 13  sales_lag1          15567 non-null  float64
 14  sales_lag2          15567 non-null  float64
 15  month_sin           15567 non-null  float64
 16  mont

In [ ]:
"""
# Encode article_name
ordinal_encoder = OrdinalEncoder()
df['article_name'] = ordinal_encoder.fit_transform(df[['article_name']])
"""

"\n# Encode article_name\nordinal_encoder = OrdinalEncoder()\ndf['article_name'] = ordinal_encoder.fit_transform(df[['article_name']])\n"

In [ ]:
#df.isnull().sum()

In [ ]:
"""
# Fill lag1_sales_7d_mean with 0
df['lag1_sales_7d_mean'].fillna(0, inplace=True)

# Fill lag2_sales_30d_std with mean or median
mean = df['lag2_sales_30d_std'].mean()
median = df['lag2_sales_30d_std'].median()

if len(df) > 500:
  df['lag2_sales_30d_std'].fillna(mean, inplace=True)
else:
  df['lag2_sales_30d_std'].fillna(median, inplace=True)
  """

"\n# Fill lag1_sales_7d_mean with 0\ndf['lag1_sales_7d_mean'].fillna(0, inplace=True)\n\n# Fill lag2_sales_30d_std with mean or median\nmean = df['lag2_sales_30d_std'].mean()\nmedian = df['lag2_sales_30d_std'].median()\n\nif len(df) > 500:\n  df['lag2_sales_30d_std'].fillna(mean, inplace=True)\nelse:\n  df['lag2_sales_30d_std'].fillna(median, inplace=True)\n  "

In [ ]:
#Fdf = df.drop(['promoted', 'yearweek_end', 'day_of_week', 'day_of_week_sin', 'day_of_week_cos', 'day_promo02'], axis=1)

In [ ]:
# Regression models

X = df.drop(['sales'], axis=1)
y = df['sales']

rf = RandomForestRegressor()
rf.fit(X, y)

RandomForestRegressor()

In [ ]:
start=len(df)

In [ ]:
end=len(df)+4

In [ ]:
# Make predictions
X_test = X.iloc[-4:]
rf_predictions = rf.predict(X_test)

# Calculate error metrics
rf_mse = mean_squared_error(y[-4:], rf_predictions)
rf_mape = mean_absolute_percentage_error(y[-4:], rf_predictions)

In [ ]:
print(rf_predictions, rf_mse, rf_mape)

[10762.11 11653.4  11653.4  11653.4 ] 0.0030250000000320142 2.5552871221859076e-06


In [ ]:
test = pd.read_csv('/content/drive/MyDrive/ca_training_data_hs23/merged_test_data.csv')

In [ ]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   article_name    180 non-null    object
 1   date_id         180 non-null    int64 
 2   year            180 non-null    int64 
 3   yearweek_start  180 non-null    object
 4   yearweek_end    180 non-null    object
 5   month           180 non-null    int64 
 6   week            180 non-null    int64 
 7   boss_thema_id   180 non-null    int64 
 8   promo_01        180 non-null    int64 
 9   promo_02        180 non-null    int64 
 10  promo_03        180 non-null    int64 
 11  promo_04        180 non-null    int64 
 12  promo_05        180 non-null    int64 
dtypes: int64(10), object(3)
memory usage: 18.4+ KB


In [ ]:
# Regression models test

X_n = df['article_name', 'date_id', 'year', 'month', 'week', 'promo_01']
y_n = df['sales']

r = RandomForestRegressor()
r.fit(X_n, y_n)

KeyError: ignored

In [ ]:
# Make predictions
Xtest = X_n.iloc[-4:]
rfpredictions = r.predict(Xtest)

# Calculate error metrics
r_mse = mean_squared_error(y[-4:], rfpredictions)
r_mape = mean_absolute_percentage_error(y[-4:], rfpredictions)

In [ ]:
N_predictions = rf.predict(test)

#YAYYYYYYYYY REALLY GOOD RESULTS for a regression model.


- MSE of 0.16 is quite low. This means the model predictions are very close to the actual values on average. Lower is better for MSE.

- MAPE of 1.8583906337914633e-05 is extremely low(approximately 0.0019%). This is the average percent error between predictions and actuals. You generally want MAPE to be as low as possible. Below 10% is excellent, and under 5% is superb for most problems.

- The very small MAPE indicates the percent errors are tiny on average.

- Our model is predicting the target variable very accurately. The errors are within a fraction of a percent of the true values on average.

###Next steps:
Check performance on test set too.

In [ ]:
# Convert Series to numpy array
#predictions = rf_predictions.values


print(rf_predictions.shape)

pred_df = pd.DataFrame(rf_predictions, columns=['predictions(RFR)'])

output_path = '/content/drive/MyDrive/ca_training_data_hs23'

pred_df.to_csv(os.path.join(output_path,'regressor_predictions.csv'), index=False)


(4,)


In [ ]:
import pickle
# Save model
with open('regressorModel.pkl', 'wb') as file:
    pickle.dump(rf, file)


In [ ]:
# Load model
with open('regressorModel.pkl', 'rb') as file:
   loaded_model = pickle.load(file)

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/ca_training_data_hs23/final_feat.csv')

df.info()

In [ ]:
"""

# Make predictions
X_test = X.iloc[-4:]
rf_predictions = rf.predict(X_test)

# Calculate error metrics
rf_mse = mean_squared_error(y[-4:], rf_predictions)
rf_mape = mean_absolute_percentage_error(y[-4:], rf_predictions)
print(rf_predictions, rf_mse, rf_mape)

"""

'\n\n# Make predictions\nX_test = X.iloc[-4:]\nrf_predictions = rf.predict(X_test)\n\n# Calculate error metrics\nrf_mse = mean_squared_error(y[-4:], rf_predictions)\nrf_mape = mean_absolute_percentage_error(y[-4:], rf_predictions)\nprint(rf_predictions, rf_mse, rf_mape)\n\n'

In [ ]:
"""
output_path = '/content/drive/MyDrive/ca_training_data_hs23'
Fdf.to_csv(os.path.join(output_path, 'final_features.csv'), index=False)
"""

"\noutput_path = '/content/drive/MyDrive/ca_training_data_hs23'\nFdf.to_csv(os.path.join(output_path, 'final_features.csv'), index=False)\n"


- The predictions are still very similar, with 3 values clustered around 11,653 and 1 value at 10,760.

- A value of 0.92 indicates the predictions are much closer to the true values on average.


MAPE: 0.00004460137520989011

- The MAPE is still extremely small, around 0.004%

- This means the percent error is very low.


In [ ]:
"""
# Model Optimization(Optional since we have really good results)

# Random Forest feature importance
rf.feature_importances_

# Remove least important features
X_reduced = X.drop(X.columns[rf.feature_importances_.argsort()[:3]], axis=1)

# Retrain model
rf_reduced = RandomForestRegressor()
rf_reduced.fit(X_reduced, y)


# Final Forecasts

# Retrain final RF model
rf_final = rf_reduced
rf_forecasts = rf_final.predict(X_test)

print(rf_forecasts)
"""

'\n# Model Optimization(Optional since we have really good results)\n\n# Random Forest feature importance\nrf.feature_importances_\n\n# Remove least important features\nX_reduced = X.drop(X.columns[rf.feature_importances_.argsort()[:3]], axis=1)\n\n# Retrain model\nrf_reduced = RandomForestRegressor()\nrf_reduced.fit(X_reduced, y)\n\n\n# Final Forecasts\n\n# Retrain final RF model\nrf_final = rf_reduced\nrf_forecasts = rf_final.predict(X_test)\n\nprint(rf_forecasts)\n'

This is just in case the accuracy reduces when we are given the test set